File: 155-min-stack/155-min-stack.java
class MinStack {
    class Node {
        int val;
        int min;
        Node next;
        Node(int val, int min, Node node) {
            this.val = val;
            this.min = min;
            this.next = node;
        }
    }
    Node curr;
    public MinStack() {
    }
    
    public void push(int val) {
        if(curr == null) 
            curr = new Node(val, val, null);
        else {
            Node node = new Node(val, Math.min(curr.min, val), curr);
            curr = node;
        }
    }
    
    public void pop() {
        curr = curr.next;
    }
    
    public int top() {
        return curr.val;
    }
    
    public int getMin() {
        return curr.min;
    }
}

File: 155-min-stack/155-min-stack.py
class MinStack:
    def __init__(self):
        self.arr = []
        self.ans = []

    def push(self, val: int) -> None:
        self.arr.append(val)
        if len(self.ans) > 0:
            self.ans.append(min(val, self.ans[-1]))
        else:
            self.ans.append(val)

    def pop(self) -> None:
            self.arr.pop()
            self.ans.pop()

    def top(self) -> int:
        return self.arr[-1]

    def getMin(self) -> int:
        return self.ans[-1]